@here/harp-mapview/lib/text/LayoutState.ts

```diff
@@ -0,0 +1,137 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+    DefaultTextStyle,
+    HorizontalAlignment,
+    TextLayoutStyle,
+    VerticalAlignment
+} from "@here/harp-text-canvas";
+
+/**
+ * Version of layout style stored.
+ */
+export enum LayoutUsed {
+    Undefined = 0,
+    ThemeBased,
+    Alternative
+}
+
+/**
+ * Defines possible text placement relative to anchor.
+ */
+export interface AnchorPlacement {
+    h: HorizontalAlignment;
+    v: VerticalAlignment;
+}
+
+/**
+ * Layout state of the text part of the `TextElement`.
+ *
+ * Used mainly for multi-anchor placement algorithm.
+ * @hidden
+ */
+export class LayoutState {
+    private m_state = LayoutUsed.Undefined;
+    private m_hAlign = DefaultTextStyle.DEFAULT_HORIZONTAL_ALIGNMENT;
+    private m_vAlign = DefaultTextStyle.DEFAULT_VERTICAL_ALIGNMENT;
+
+    /**
+     * Set layout based on theme style defined and optional text placement.
+     *
+     * @param layoutStyle The basic theme label style.
+     * @param placement The optional new anchor placement.
+     */
+    setPlacement(layoutStyle: TextLayoutStyle, placement?: AnchorPlacement) {
+        if (placement !== undefined) {
+            this.m_hAlign = placement.h;
+            this.m_vAlign = placement.v;
+            this.m_state =
+                placement.h === layoutStyle.horizontalAlignment &&
+                placement.v === layoutStyle.verticalAlignment
+                    ? LayoutUsed.ThemeBased
+                    : LayoutUsed.Alternative;
+        } else {
+            this.m_hAlign = layoutStyle.horizontalAlignment;
+            this.m_vAlign = layoutStyle.verticalAlignment;
+            this.m_state = LayoutUsed.ThemeBased;
+        }
+    }
+
+    /**
+     * Acquire current placement setup.
+     *
+     * Function returns alternative or base placement depending on layout state.
+     *
+     * @param layoutStyle The label layout style.
+     * @returns The current anchor placement.
+     */
+    getPlacement(layoutStyle: TextLayoutStyle): AnchorPlacement {
+        if (this.isAlternative()) {
+            return { h: this.m_hAlign, v: this.m_vAlign };
+        } else {
+            return { h: layoutStyle.horizontalAlignment, v: layoutStyle.verticalAlignment };
+        }
+    }
+
+    /**
+     * Reset existing `LayoutState` to appear like a fresh state.
+     */
+    reset() {
+        this.m_state = LayoutUsed.Undefined;
+    }
+
+    /**
+     * @returns `true` if element has not yet been placed and layout variant assigned.
+     */
+    isUndefined(): boolean {
+        return this.m_state === LayoutUsed.Undefined;
+    }
+
+    /**
+     * @returns `true` if element is using original style defined in theme.
+     */
+    isBase(): boolean {
+        return this.m_state === LayoutUsed.ThemeBased;
+    }
+
+    /**
+     * @returns `true` if element is fading in.
+     */
+    isAlternative(): boolean {
+        return this.m_state === LayoutUsed.Alternative;
+    }
```

Comment:
I don't understand why you need this state.
You can just save here whatever was the last alignment used.

Reply:
It is important to know because of label.bounds being cached from the last frame. Imagine two scenario:
- label bounds calculated for alternative placement, then got invisible/rejected, label placed again with bounds already stored (in label.bounds property) -> we should calculate placement for base layout but we have bounds there for alternative placement -> bounds should be re-calculated.
- secondly we need to know if the LayoutState is storing any reasonable values before using them.

Reply:
Ok, just to limit discussions, next update includes:
- removed enum state inside LayoutState,
- layout state created seldomly only when really needed.